src/zip_cycle.rs

```diff
@@ -0,0 +1,119 @@
+use super::size_hint;
+
+// ZipCycle originally written by Aistis Raulinaitis
+
+/// An iterator which iterates two other iterators simultaneously
+/// and cycles the shorter iter until the longer is finished.
+///
+/// See [`.zip_cycle()`](../trait.Itertools.html#method.zip_cycle) for more information.
+#[derive(Debug, PartialEq, Clone)]
+pub struct ZipCycle<A, B>
+where
+    A: Clone + Iterator,
+    B: Clone + Iterator,
+{
+    ab: Option<AB>,
+    a_orig: A,
+    a: A,
+    b_orig: B,
+    b: B,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+enum AB {
+    A,
+    B,
+}
+
+/// A `ZipCycle` on construction with `ZipCycle::new()` may end up in 3 error states, all having
+/// to do with either one or both of the underlying iterators being empty.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum ZipCycleErr {
+    /// The case both are empty
+    BothItersEmpty,
+    /// The case the left hand side is empty
+    FirstIterEmpty,
+    /// The case the right hand side is empty
+    SecondIterEmpty
+}
+
+pub fn zip_cycle<A, B>(a: A, b: B) -> Result<ZipCycle<A, B>, ZipCycleErr>
+where
+    A: Clone + Iterator,
+    B: Clone + Iterator,
+{
+    match (a.clone().peekable().peek(), b.clone().peekable().peek()) {
```

Comment:
Is it really necessary to make a `peekable` here? Couldn't we just check `a.clone().next()`?